File: swissknife/groups.py
from typing import Any, Dict, List

from .graph_client import GraphClient


def list_groups(client: GraphClient, top: int = 25) -> List[Dict[str, Any]]:
    """
    Список первых top групп (AAD / M365 Groups / Teams backing groups).
    """
    params = {"$top": top}
    result = client.get("/groups", params=params)
    return result.get("value", [])


def get_group(client: GraphClient, group_id: str) -> Dict[str, Any]:
    """
    Получить одну группу по id.
    """
    path = f"/groups/{group_id}"
    return client.get(path)


def list_group_members(client: GraphClient, group_id: str) -> List[Dict[str, Any]]:
    """
    Список участников группы (users / groups / service principals).
    """
    path = f"/groups/{group_id}/members"
    result = client.get(path)
    return result.get("value", [])


def add_group_owner_by_upn(
    client: GraphClient,
    group_id: str,
    user_upn: str,
) -> Dict[str, Any]:
    """
    Добавить owner'а в группу по UPN.

    1) Получаем пользователя по /users/{user_upn}
    2) Добавляем в /groups/{group_id}/owners/$ref

    Требуются права: Group.ReadWrite.All + User.Read.All / Directory.Read.All
    """
    user = client.get(f"/users/{user_upn}")
    user_id = user.get("id")
    if not user_id:
        raise RuntimeError(f"Не найден пользователь {user_upn}")

    body = {
        "@odata.id": f"https://graph.microsoft.com/v1.0/users/{user_id}"
    }

    path = f"/groups/{group_id}/owners/$ref"
    result = client.post(path, json=body)
    return result

File: app/internal/services/groups.go
package services

import (
	"encoding/json"
	"net/url"

	"swissknife-app/internal/session"
)

type GroupsService struct {
	s *session.Session
}

func NewGroupsService(s *session.Session) *GroupsService { return &GroupsService{s: s} }

func (g *GroupsService) List(search string, maxItems int) ([]json.RawMessage, error) {
	c, err := g.s.Client()
	if err != nil {
		return nil, err
	}
	params := url.Values{"$top": {"100"}}
	if search != "" {
		params.Set("$filter", "startswith(displayName,'"+escapeODataLiteral(search)+"')")
	}
	return c.ListAll(g.s.Ctx(), "/groups", params, maxItems)
}

func (g *GroupsService) Get(groupID string) (json.RawMessage, error) {
	c, err := g.s.Client()
	if err != nil {
		return nil, err
	}
	var out json.RawMessage
	err = c.Get(g.s.Ctx(), "/groups/"+url.PathEscape(groupID), nil, &out)
	return out, err
}

func (g *GroupsService) Members(groupID string) ([]json.RawMessage, error) {
	c, err := g.s.Client()
	if err != nil {
		return nil, err
	}
	return c.ListAll(g.s.Ctx(), "/groups/"+url.PathEscape(groupID)+"/members", nil, 0)
}

func (g *GroupsService) addRef(groupID, upn, kind string) error {
	if err := g.s.GuardWrite(); err != nil {
		return err
	}
	c, err := g.s.Client()
	if err != nil {
		return err
	}
	// first resolve the UPN into an id
	var user struct {
		ID string `json:"id"`
	}
	if err := c.Get(g.s.Ctx(), "/users/"+url.PathEscape(upn), url.Values{"$select": {"id"}}, &user); err != nil {
		return err
	}
	body := map[string]any{"@odata.id": "https://graph.microsoft.com/v1.0/users/" + user.ID}
	err = c.Post(g.s.Ctx(), "/groups/"+url.PathEscape(groupID)+"/"+kind+"/$ref", body, nil)
	g.s.Record("groups.add_"+kind, groupID, "user="+upn, err)
	return err
}

func (g *GroupsService) AddOwner(groupID, upn string) error {
	return g.addRef(groupID, upn, "owners")
}

func (g *GroupsService) AddMember(groupID, upn string) error {
	return g.addRef(groupID, upn, "members")
}

// CreateM365 creates a Unified group; ownerUpn is optional.
func (g *GroupsService) CreateM365(displayName, description, mailNickname, ownerUpn string) (json.RawMessage, error) {
	if err := g.s.GuardWrite(); err != nil {
		return nil, err
	}
	c, err := g.s.Client()
	if err != nil {
		return nil, err
	}
	body := map[string]any{
		"displayName":     displayName,
		"description":     description,
		"groupTypes":      []string{"Unified"},
		"mailEnabled":     true,
		"securityEnabled": false,
		"mailNickname":    mailNickname,
	}
	if ownerUpn != "" {
		userURL := "https://graph.microsoft.com/v1.0/users('" + ownerUpn + "')"
		body["owners@odata.bind"] = []string{userURL}
		body["members@odata.bind"] = []string{userURL}
	}
	var out json.RawMessage
	err = c.Post(g.s.Ctx(), "/groups", body, &out)
	g.s.Record("groups.create", displayName, "nickname="+mailNickname, err)
	return out, err
}
